source/core_sntp_serializer.c

```diff
@@ -259,50 +296,104 @@ static SntpStatus_t calculateClockOffset( const SntpTimestamp_t * pClientTxTime,
+         * of each other, we will prepare for the clock-offset calculation. To calculate
+         * clock-offset, the first order difference values need to be stored as signed integers
+         * in the correct polarity of differences according to the formula. */
+        int32_t signedFirstOrderDiffSend = 0;
+        int32_t signedFirstOrderDiffRecv = 0;
+
+        /* To avoid overflow issues in assigning the unsigned values of first order differences
+         * to signed integers, we will inverse the values if they represent a large value that cannot be
+         * represented in a signed integer. */
+        if( ( firstOrderDiffSend & UINT32_MOST_SIGNIFICANT_BIT_BIT_MASK )
+            == UINT32_MOST_SIGNIFICANT_BIT_BIT_MASK )
 
-        /* Perform ( T2 - T1 ) offset calculation of SNTP Request packet path. */
-        firstOrderDiffSend = pServerRxTime->seconds - pClientTxTime->seconds;
+        {
+            /* Negate the unsigned integer by performing 2's complement operation.
+             * Note: This is done to be compliant with both CBMC and MISRA Rule 10.1.
+             * CBMC flags overflow for (unsigned int = 0U - positive value) whereas
+             * MISRA rule forbids use of unary minus operator on unsigned integers. */
+            firstOrderDiffSend = ~firstOrderDiffSend + 1U;
+            sendPolarity = !sendPolarity;
+        }
+
+        if( ( firstOrderDiffRecv & UINT32_MOST_SIGNIFICANT_BIT_BIT_MASK )
+            == UINT32_MOST_SIGNIFICANT_BIT_BIT_MASK )
+
+        {
+            /* Negate the unsigned integer by performing 2's complement operation.
+             * Note: This is done to be compliant with both CBMC and MISRA Rule 10.1.
+             * CBMC flags overflow for (unsigned int = 0U - positive value) whereas
+             * MISRA rule forbids use of unary minus operator on unsigned integers. */
+            firstOrderDiffRecv = ~firstOrderDiffRecv + 1U;
+            recvPolarity = !recvPolarity;
+        }
+
+        /* Now we can safely store the first order differences as signed integer values. */
+        signedFirstOrderDiffSend = ( int32_t ) firstOrderDiffSend;
+        signedFirstOrderDiffRecv = ( int32_t ) firstOrderDiffRecv;
 
-        /* Perform ( T3 - T4 ) offset calculation of SNTP Response packet path. */
-        firstOrderDiffRecv = 0U - firstOrderDiff;
+        /* To calculate the clock-offset value, we will correct the signed first order difference
+        * values to match the subtraction polarity direction of "Server Time" - "Client Time". */
+        if( sendPolarity == false )
+        {
+            signedFirstOrderDiffSend = 0 - signedFirstOrderDiffSend;
+        }
 
-        /* Perform second order calculation of using average of the above offsets. */
-        sumOfFirstOrderDiffs = firstOrderDiffSend + firstOrderDiffRecv;
+        if( recvPolarity == false )
+        {
+            signedFirstOrderDiffRecv = 0 - signedFirstOrderDiffRecv;
```

Comment:
CBMC reports this as a signed int overflow during subtraction.

This should be easy to fix though: `-signedFirstOrderDiffSend` should make both CBMC and MISRA happy.